### PR TITLE
core(driver): create eval code using interface

### PR DESCRIPTION
**Pull request by connorjclark**
`Idea 1` from #10781

Where simple, I also converted from `getElementsInDocumentString` to `getElementsInDocument`. Skipped for tap-targets, was too much.

Copied `HTMLElementTagNameMap` from `dom.js` for `getElementsInDocument`. I renamed it because TS wasn't picking it up, I'm surprised it even worked in `dom.js` because it's overriding the same identifier.

### CODE REVIEW

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:440]: Thanks for the great feedback!

> js-style "overloading" tends to grow in implementation complications over time and is harder to understand as a caller (e.g. options could be included with a string expression, but would it do anything in that case? Now we have to document that, etc)

I took a second pass to incorporate this feedback, and added some docs. Running to a meeting, will respond to the rest later.

**Comment by connorjclark** [comment]
> Isn't it called HTMLElementByTagName though? I don't think it should be overriding anything..

Yeah, I messed up the copy/paste and got confused. Put the names back. Should one of these files do a type import for the other?

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [inline lighthouse-core/gather/driver.js:428]: Could drop this. if we like this, I'd follow up with a PR that changes all existing usages to this.

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [inline lighthouse-core/gather/driver.js:439]: FYI I would like to drop the `string` part eventually.

**Comment by brendankenny** [comment]
> Should one of these files do a type import for the other?

I had the same thought but the files are in such different contexts I don't know which one should import from the other :)

File: lighthouse-core/gather/driver.js
Comment (brendankenny) [reply lighthouse-core/gather/driver.js:439]: > FYI I would like to drop the `string` part eventually.

could we drop it now and say "if you still want to use a string use the legacy `evaluateAsync()`"?

Comment (connorjclark) [reply lighthouse-core/gather/driver.js]: done

**Comment by connorjclark** [comment]
@brendankenny any more feedback?

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:440]: Would you prefer splitting this out to `evaluate(fn, ...)` and `evaluateString(code, ...)` (alias for `evalauteAsync`)?

File: lighthouse-core/gather/driver.js
Comment (brendankenny) [inline lighthouse-core/gather/driver.js:461]: possible to split this into a separate PR since it's mostly an independent change?

File: lighthouse-core/gather/driver.js
Comment (brendankenny) [reply lighthouse-core/gather/driver.js:440]: rather than add another `evaluate*()` hop with `_evaluate`, what do you think about keeping it called `evaluateAsync`, so

`evaluate(fn, ...)` calls `evaluateAsync(string, opts)` calls `_evaluateInContext(string, contextId)`

then in the next breaking change we either incorporate the guts of `evaluateAsync` into `evaluate` so it's back to only two functions, or we rename `evaluateAsync` to `evaluateString` or whatever if we do want to keep it around for folks to use?

I see the reason for doing it the `_evaluate` way, but the number of evaluate layers doesn't seem worth it when external to driver we still only have `evaluate()` and `evaluateAsync` either way.

File: lighthouse-core/lib/page-functions.js
Comment (brendankenny) [inline lighthouse-core/lib/page-functions.js:23]: could this be directly incorporated into `evaluate()` (and `evaluateFunctionOnObject`) instead of splitting it into this file? It's mostly not shared code since the two callers take different branches in the conditional (and it's not really a "page function" in the same way as the other functions in this file are).

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:461]: will do

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:440]: done

File: lighthouse-core/lib/page-functions.js
Comment (connorjclark) [reply lighthouse-core/lib/page-functions.js:23]: I prefer it here bc/ it's easier to test (driver has some overhead with `beforeEach` etc.) and 
keeps the driver file smaller (it's so big). Maybe `lib/eval.js`?

And then export two function: `createFunctionEvalCode` and `createIffeEvalCode`?

**Comment by paulirish** [comment]
We think this is waiting on another review @brendankenny

File: lighthouse-core/gather/driver.js
Comment (adamraine) [reply lighthouse-core/gather/driver.js:457]: I'm a little confused about the purpose of `_evaluate`. Is it for a possible patch where `evaluateAsync` is removed but we still need a function for executing short expressions?

Comment (connorjclark) [reply lighthouse-core/gather/driver.js:476]: I think I just forgot to delete `_evaluate`.

**Comment by adamraine** [comment]
@connorjclark is next action on reviewer?

File: lighthouse-core/lib/page-functions.js
Comment (adamraine) [inline lighthouse-core/lib/page-functions.js:25]: nit: `join()` on an empty list is an empty string anyway. You can get rid of the conditionals.

File: lighthouse-core/test/lib/page-functions-test.js
Comment (adamraine) [inline lighthouse-core/test/lib/page-functions-test.js:96]: nit: add expectation for `eval(code)`

**Review by adamraine — CHANGES_REQUESTED** [review]
Big fan of these changes. Is the plan still to switch most instances of `evaluateAsync` to use `evaluate` in a follow-up PR?

**Comment by connorjclark** [comment]
> Big fan of these changes. Is the plan still to switch most instances of evaluateAsync to use evaluate in a follow-up PR?

We could do that change now–evaluate is a drop in replacement. I think we need to keep `evaluateAsync` just for plugins that happen to use it (are there any?). It's easy to keep around.

There is no technical reason for two functions, it's just that we now would prefer the name `evaluate` (we don't name all of our functions `*Async` if it returns a promise...)

File: lighthouse-core/lib/page-functions.js
Comment (connorjclark) [reply lighthouse-core/lib/page-functions.js:25]: args and deps are optional params.

strangely, typecheck found no error when I removed the conditionals, even though both were marked as `Array | undefined` ...

**Comment by adamraine** [comment]
> We could do that change now–evaluate is a drop in replacement

SGTM

**Comment by connorjclark** [comment]
Wait, I just remembered why I didn't want to do that in this PR. It's a lot of small, sometimes tricky changes. I'll do in a followup.

File: lighthouse-core/gather/gatherers/script-elements.js
Comment (adamraine) [inline lighthouse-core/gather/gatherers/script-elements.js:11]: Should we keep this consistent with the style of link-elements?

**Review by adamraine — APPROVED** [review]
Overall LGTM

File: lighthouse-core/gather/driver.js
Comment (brendankenny) [inline lighthouse-core/gather/driver.js:446]: can delete

File: lighthouse-core/gather/driver.js
Comment (brendankenny) [inline lighthouse-core/gather/driver.js:471]: > see `createEvalCode`

this is the interface exposed to gathering. Are there any docs on `createEvalCode` that could move here?

File: lighthouse-core/gather/driver.js
Comment (brendankenny) [inline lighthouse-core/gather/driver.js:475]: I think this will have to be tweaked to get the args right (it should end up being a tuple). e.g. this would allow an array of strings to `fetcher`'s `injectIframe`, not a single string, and if the parameters are of mixed types I guess this would end up with an array of the union of the types?

File: lighthouse-core/gather/gatherers/script-elements.js
Comment (brendankenny) [reply lighthouse-core/gather/gatherers/script-elements.js]: > Should we keep this consistent with the style of link-elements?

yeah, which is the way that breaks when minified? If we're getting confused we should definitely document this :)

File: lighthouse-core/lib/page-functions.js
Comment (brendankenny) [reply lighthouse-core/lib/page-functions.js:25]: > strangely, typecheck found no error when I removed the conditionals, even though both were marked as `Array | undefined` ...

it's because this file is `// @ts-nocheck`. Maybe after this change and followup(s), we could drop that with a minimum of expect-errors

File: lighthouse-core/gather/driver.js
Comment (brendankenny) [inline lighthouse-core/gather/driver.js:452]: revert?

File: lighthouse-core/gather/gatherers/script-elements.js
Comment (connorjclark) [reply lighthouse-core/gather/gatherers/script-elements.js:11]: Yes, and it is necessary, due to mangling #11463

File: lighthouse-core/lib/page-functions.js
Comment (brendankenny) [reply lighthouse-core/lib/page-functions.js:23]: > I prefer it here bc/ it's easier to test (driver has some overhead with `beforeEach` etc.) and
> keeps the driver file smaller (it's so big). Maybe `lib/eval.js`?
> 
> And then export two function: `createFunctionEvalCode` and `createIffeEvalCode`?

well it would definitely be good to split the two, but it'll also have a new home after #11633, and it also looks trivial to add to `.evaluateAsync` in `driver-test`? 

```
const {expression} = connectionStub.sendCommand.findInvocation('Runtime.evaluate');
expect(expression).toEqual(`(() => {
function mainFn() {
  return true;
}
return mainFn();
})()`);
```

File: lighthouse-core/test/lib/page-functions-test.js
Comment (brendankenny) [inline lighthouse-core/test/lib/page-functions-test.js:23]: to channel my inner @patrickhulce, since these are static, should just use `expect(code).toEqual(...)`

File: lighthouse-core/gather/gatherers/script-elements.js
Comment (brendankenny) [reply lighthouse-core/gather/gatherers/script-elements.js]: > Yes, and it is necessary, due to mangling #11463

this is unfortunate, since we lose all `deps` type checking benefits then :/

(see also https://github.com/GoogleChrome/lighthouse/pull/11463#discussion_r497088597 for an example of the mangling)

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:475]: tried this but didnt work

image

image


then I found this: https://github.com/microsoft/TypeScript/issues/1024#issuecomment-343707528

Still parsing how this relates to variadic types / if this is now possible in typescript ...

Comment (connorjclark) [reply lighthouse-core/gather/driver.js:475]: doh! just needed to change `args?: T[]` to `args?: T`

File: lighthouse-core/lib/page-functions.js
Comment (connorjclark) [reply lighthouse-core/lib/page-functions.js:27]: @patrickhulce would you prefer this going into `driver.js` or a new file `lib/eval.js`?

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:451]: weird, it's at 101 columns but yarn lint doesn't flag it right now..

File: lighthouse-core/lib/page-functions.js
Comment (patrickhulce) [reply lighthouse-core/lib/page-functions.js:27]: new `lib/eval.js` file SGTM, though I'll also +1 @brendankenny's test suggest to make sure `.evaluate` uses it :)

File: lighthouse-core/test/lib/page-functions-test.js
Comment (patrickhulce) [reply lighthouse-core/test/lib/page-functions-test.js]: I will say on the spectrum of obvious `toMatchInlineSnapshot` to obvious `toEqual`, this particular test does feel somewhat ambiguous to me (specific whitespace seems likely to change with no correct answer, but other important bits never should for example), but I love the reinforcement of the rule so much I'll go along with it 😆

File: lighthouse-core/lib/page-functions.js
Comment (brendankenny) [reply lighthouse-core/lib/page-functions.js:27]: > new lib/eval.js file

Isn't the (currently named) `RuntimeController` going to be this already in #11633, though? I guess I don't understand what the problem is just inlining the five lines or so :)

Comment (patrickhulce) [reply lighthouse-core/lib/page-functions.js:27]: > Isn't the (currently named) RuntimeController going to be this already in #11633, though?

Ya, that's why I don't really care much where it lives in this PR :) 

> I guess I don't understand what the problem is just inlining the five lines or so

I mean I don't feel terribly strongly about it, the rebase gets slightly more annoying with inline but not by much.

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:473]: No. But we could copy the contents of createEvalCode here?

I'm surprised `@see` doesn't do anything useful in vscode.

File: lighthouse-core/gather/gatherers/script-elements.js
Comment (connorjclark) [reply lighthouse-core/gather/gatherers/script-elements.js]: yea, reminder tho that we could get it  back if we namespace the page functions (idea 3 https://github.com/GoogleChrome/lighthouse/pull/10781 )

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:474]: ref https://github.com/microsoft/TypeScript/issues/35524

**Comment by connorjclark** [comment]
thanks for the review @adamraine 

@brendankenny is there anything else you'd like to see changed?

File: lighthouse-core/gather/driver.js
Comment (brendankenny) [inline lighthouse-core/gather/driver.js:491]: ok, last pitch on this :) Having it inline means no (yet another) evaluateAsync helper method to look through to see how the whole system works, no `mode`, and no duplication of the relatively complicated types here. e.g.

```js
  /**
   * Evaluate a function in the context of the current page. If `useIsolation` is true, the function
   * will be evaluated in a content script that has access to the page's DOM but whose JavaScript state
   * is completely separate.
   * Returns a promise that resolves on a value of `mainFn`'s return type.
   * @template {any[]} T, R
   * @param {(...args: T) => R} mainFn The main function to call.
   * @param {{useIsolation?: boolean, args?: T, deps?: Array<Function|string>}=} options `args` should
   *   match the args of `mainFn`, and can be any serializable value. `deps` are functions that must be
   *   defined for `mainFn` to work.
   * @return {Promise<R>}
   */
  async evaluate(mainFn, options = {}) {
    const argsSerialized = options.args ? options.args.map(a => JSON.stringify(a)).join(',') : '';
    const depsSerialized = options.deps ? options.deps.join('\n') : '';

    const expression = `(() => {
      ${depsSerialized}
      ${mainFn}
      return ${mainFn.name}(${argsSerialized});
    })()`;

    return this.evaluateAsync(expression, options);
  }
```
(docs stolen from `evaluateAsync` and `createEvalCode`)

Comment (brendankenny) [reply lighthouse-core/gather/driver.js:491]: interestingly, since `args` is optional, a caller can leave them out even if required and be ok with the type checker. I'm not sure if there's a way around that except
- requiring `options`/`args` (so all callers would have to at least provide an empty `[]` when no args are needed)
- or do it with rest params like how we do `sendCommand`, and we know the pain that causes

File: lighthouse-core/test/gather/driver-test.js
Comment (connorjclark) [reply lighthouse-core/test/gather/driver-test.js:323]: I had trouble getting a literal string to work here (was always off, and it being a substring check I couldn't tell how), so I gave up and used what you see now. Will try again.

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:470]: Was under false impression typedef was necessary. fixed.

File: lighthouse-core/gather/driver.js
Comment (connorjclark) [reply lighthouse-core/gather/driver.js:491]: I'm resisting splitting this up because it's harder to test through runtime.evaluate sniffing than directly on the string generated.

> interestingly, since args is optional ...

I tried making a conditional type that checks for tuple length and returns `{args?: T} : {}` or some variation, and that part worked fine enough, but getting the default values for the options params to work proved difficult, so I gave up and just went with requiring an empty args array.

Comment (brendankenny) [reply lighthouse-core/gather/driver.js]: > I'm resisting inlining this because it's harder to test through runtime.evaluate sniffing than directly on the string generated.

But it's the same string, just surrounded by `_evaluateInContext` pre- and post-text :) (and if we replaced `window` in `_evaluateInContext` with `globalThis`, we could even `eval()` that code too).

Comment (brendankenny) [reply lighthouse-core/gather/driver.js]: > but getting the default values for the options params to work proved difficult

yes! it gets weird with a `never` appearing and really mucking up the works

**Comment by connorjclark** [comment]
saving for separate pr

```
return `function () {
  ${depsSerialized}
  ${mainFn}
  return ${mainFn.name}.call(this, ${argsSerialized});
}`;

----

it('function', () => {
  function mainFn(a, b) {
    return adder(a, b);
  }
  function adder(a, b) {
    return a + b;
  }
  const code = createEvalCode(mainFn, {
    args: [1, 2],
    mode: 'function',
    deps: [adder],
  });
  expect(code).toEqual(`function () {
    function adder(a, b) {
    return a + b;
  }
    function mainFn(a, b) {
    return adder(a, b);
  }
    return mainFn.call(this, 1,2);
  }`);
  expect(eval(`(${code})()`)).toEqual(3);
});
```

File: lighthouse-core/test/gather/driver-test.js
Comment (brendankenny) [inline lighthouse-core/test/gather/driver-test.js:348]: with this `eval` idea you're introducing we could probably add some tests that exercise the error catching some time

cc @patrickhulce :)

File: lighthouse-core/test/gather/driver-test.js
Comment (brendankenny) [inline lighthouse-core/test/gather/driver-test.js:352]: not essential, but maybe add a comment that this is mocking this to skip the promise/error-catching wrapper?

File: lighthouse-core/test/gather/driver-test.js
Comment (brendankenny) [inline lighthouse-core/test/gather/driver-test.js:378]: no need to change, but I'm curious if there's a reason to make this `any`?

File: lighthouse-core/test/gather/driver-test.js
Comment (brendankenny) [inline lighthouse-core/test/gather/driver-test.js:396]: possibly silly overkill idea: since this file is type checked, this could also serve a test for the return type as well (even though the return _value_ will be meaningless in this case). Something like

```suggestion
    /** @type {{a: number, b: number, passThru: any}} */
    const result = await driver.evaluate(mainFn, { // eslint-disable-line no-unused-vars
```

**Review by brendankenny — APPROVED** [review]
LGTM

File: lighthouse-core/test/gather/driver-test.js
Comment (patrickhulce) [reply lighthouse-core/test/gather/driver-test.js:348]: my hope is that *everything* in driver becomes less scary and more obvious to test well when it's not all together :)

File: lighthouse-core/test/gather/driver-test.js
Comment (connorjclark) [reply lighthouse-core/test/gather/driver-test.js:378]: nothing happens to it, so the most applicable type is any.

**Comment by connorjclark** [comment]
oof, using of globalThis requires node 12. #11656

**Comment by patrickhulce** [comment]
> well, this has been open for 544 days (I really thought I opened this much more recently...woah), so what's a few more :)

This and #11633 will have quite the merge conundrum to sort out :) Are you planning on landing this imminently?

**Comment by brendankenny** [comment]
> This and #11633 will have quite the merge conundrum to sort out :) Are you planning on landing this imminently?

shouldn't it be not too hard? It's just the one function now and the tests are a separate block (`.evaluate()` instead of `.evaluateAsync()`)

**Comment by patrickhulce** [comment]
> shouldn't it be not too hard? It's just the one function now and the tests are a separate block (.evaluate() instead of .evaluateAsync())

I mean sure, no PRs of reasonable size are ever really "too hard", but on the spectrum from normal Lighthouse PR with no merge conflicts to literally every line has a conflict, we're closer to the latter if we assume `.evaluate` logic moves into the controller as well.

**Comment by connorjclark** [comment]
Blocked on #11656 for usage of `globalThis` in our tests. Guess I'll get the honors of resolving the merge conflict.

**Comment by patrickhulce** [comment]
> Guess I'll get the honors of resolving the merge conflict.

Sorry bud 😞 

Don't worry though, I hear it's not too hard 😉

**Comment by brendankenny** [comment]
> Don't worry though, I hear it's not too hard 😉

it's not automatic, but it's not hard :P

- copy `evaluate()`
- `git merge -X theirs master`
- paste `evaluate()` back

everything is good!

- move `evaluate()` to `executionContext`

**Comment by connorjclark** [comment]
not sure what to do with the jsdocs. 100% duplicate in both places?

**Comment by brendankenny** [comment]
> not sure what to do with the jsdocs. 100% duplicate in both places?

If the implication is that the delegation pattern is awkward for humans, I agree :P

`Driver.evaluate()` is where the large majority of interaction with it will take place, so the docs should be there. But it's also strange not to document the place with the actual implementation. So, both places does seem like the best choice?

**Pull request merged**
